gobblin-data-management/src/main/java/org/apache/gobblin/util/commit/CreateDirectoryWithPermissionsCommitStep.java

```
@@ -0,0 +1,80 @@
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.commit.CommitStep;
+import org.apache.gobblin.util.HadoopUtils;
+import org.apache.gobblin.util.filesystem.OwnerAndPermission;
+
+
+/**
+ * An implementation of {@link CommitStep} for creating directories and their associated permissions before commit
+ * Necessary when creating large file paths e.g. Manifest distcp where multiple threads are creating directories at the same time,
+ * which can lead to some race conditions described in {@link org.apache.gobblin.util.HadoopUtils#unsafeRenameIfNotExists(FileSystem, Path, Path)}
+ * Current implementation only sets permissions, but it is capable of setting owner and group as well.
+ */
+@Slf4j
+public class CreateDirectoryWithPermissionsCommitStep implements CommitStep {
+  @Getter
+  Map<String, List<OwnerAndPermission>> pathAndPermissions;
+  private final URI fsUri;
+  private boolean isCompleted = false;
+
+  public CreateDirectoryWithPermissionsCommitStep(FileSystem targetFs, Map<String, List<OwnerAndPermission>> pathAndPermissions) {
+    this.pathAndPermissions = pathAndPermissions;
+    this.fsUri = targetFs.getUri();
+  }
+
+  @Override
+  public boolean isCompleted() throws IOException {
+    return isCompleted;
+  }
+
+  @Override
+  public void execute() throws IOException {
+    FileSystem fs = FileSystem.get(this.fsUri, new Configuration());
+
+    for (Map.Entry<String, List<OwnerAndPermission>> entry : pathAndPermissions.entrySet()) {
+      Path path = new Path(entry.getKey());
+      try {
+        // Is a no-op if directory already exists, stops when it hits first parent
+        // Sets the execute bit for USER in order to rename files to the folder, so it should be reset after this step is completed
+        HadoopUtils.ensureDirectoryExists(fs, path, entry.getValue().iterator());
+      } catch (AccessControlException e) {
+        log.warn("Error while setting permission on " + path, e);
+        break;
+      }
```

in the case of an error, swallow the exception and abort all further processing, yet claim `isCompleted`?